### ISSUE TITLE
Lint/NestedMethodDefinition does not detect nested define_method

### ISSUE THREAD

**Issue by madwort** [bbatsov/rubocop#2718]
Is this intentional behaviour or a missing feature?

This fails the cop:
```
def my_method
  def method_name
    do_stuff
  end
end
```

This passes the cop:
```
def my_method
  define_method(:method_name) do
    do_stuff
  end
end
```
tested with rubocop version 0.35.1

**Comment by alexdowad** [comment]
It's intentional.

**Comment by madwort** [comment]
cool, thanks.

**Comment by mgogov** [comment]
@alexdowad, out of personal interest, what's the reasoning behind this being intentional? Thanks in advance for the clarification.

**Comment by alexdowad** [comment]
> what's the reasoning behind this being intentional?

Here's my understanding of it:

Using nested `def`s is very unusual, and likely indicates that `end` keywords are mismatched somewhere, resulting in a method definition which was intended to be separate appearing inside another method definition. If `define_method` is used, it's much more likely to be deliberate.

(Of course, there might be some uses of `define_method` which are also quite unusual, and could indicate a bug.)

**Comment by mgogov** [comment]
Thanks, @alexdowad :+1:
